ledger/participant-state/kvutils/src/main/scala/com/daml/ledger/participant/state/kvutils/OffsetBuilder.scala

```diff
@@ -4,72 +4,101 @@
   *
   * @see com.daml.ledger.offset.Offset
   * @see com.daml.ledger.participant.state.kvutils.api.KeyValueParticipantStateReader
   */
 object OffsetBuilder {
-  private[kvutils] val highestStart = 0
+  val DefaultVersion: Byte = 0
+  val HighestIndexSize = 7
+
+  private[kvutils] val highestStart = 1
   private[kvutils] val middleStart = 8
   private[kvutils] val lowestStart = 12
   private[kvutils] val end = 16
 
   def onlyKeepHighestIndex(offset: Offset): Offset = {
+    val ver = version(offset)
     val highest = highestIndex(offset)
-    fromLong(highest)
+    of(ver, highest)
   }
 
   def dropLowestIndex(offset: Offset): Offset = {
-    val (highest, middle, _) = split(offset)
-    fromLong(highest, middle)
+    val (version, highest, middle, _) = split(offset)
+    of(version, highest, middle)
   }
 
   def setMiddleIndex(offset: Offset, middle: Int): Offset = {
-    val (highest, _, lowest) = split(offset)
-    fromLong(highest, middle, lowest)
+    val (version, highest, _, lowest) = split(offset)
+    of(version, highest, middle, lowest)
   }
 
   def setLowestIndex(offset: Offset, lowest: Int): Offset = {
-    val (highest, middle, _) = split(offset)
-    fromLong(highest, middle, lowest)
+    val (version, highest, middle, _) = split(offset)
+    of(version, highest, middle, lowest)
   }
 
-  def fromLong(first: Long, second: Int = 0, third: Int = 0): Offset = {
+  def of(version: Byte, first: Long, second: Int = 0, third: Int = 0): Offset = {
     val bytes = new ByteArrayOutputStream
     val stream = new DataOutputStream(bytes)
-    stream.writeLong(first)
+    stream.writeByte(version.toInt)
+    writeSevenByteLong(first, stream)
     stream.writeInt(second)
     stream.writeInt(third)
     Offset.fromByteArray(bytes.toByteArray)
   }
 
+  def version(offset: Offset): Byte = split(offset)._1
+
   // `highestIndex` is used a lot, so it's worth optimizing a little rather than reusing `split`.
   def highestIndex(offset: Offset): Long = {
     val stream = new DataInputStream(new ByteArrayInputStream(offset.toByteArray))
-    stream.readLong()
+    stream.skipBytes(1) // skip version
+    readSevenByteLong(stream)
   }
 
-  def middleIndex(offset: Offset): Int = split(offset)._2
+  def middleIndex(offset: Offset): Int = split(offset)._3
 
-  def lowestIndex(offset: Offset): Int = split(offset)._3
+  def lowestIndex(offset: Offset): Int = split(offset)._4
 
-  def split(offset: Offset): (Long, Int, Int) = {
+  def split(offset: Offset): (Byte, Long, Int, Int) = {
     val stream = new DataInputStream(new ByteArrayInputStream(offset.toByteArray))
-    val highest = stream.readLong()
+    val version = stream.readByte()
+    val highest = readSevenByteLong(stream)
     val middle = stream.readInt()
     val lowest = stream.readInt()
-    (highest, middle, lowest)
+    (version, highest, middle, lowest)
+  }
+
+  private def readSevenByteLong(stream: DataInputStream): Long = {
+    // mostly copied from DataInputStream.readLong
```

Comment:
I am against replicating this; how instead about letting `DataInputStream.readLong` read it from 8-bytes where the first one is an addition and is empty (given that the expected byte order in the buffer is big endian)?

Reply:
How would you like to provide this additional byte?